### ISSUE TITLE
[Bug] Time history output for multi-component quantities from constitutive model

### ISSUE THREAD

**Issue by francoishamon** [GEOSX/GEOSX#1552]
**Describe the bug**
I am trying to use `TimeHistory` to output a quantity (`phaseMassDensity`) living in the `DeadOilFluid` constitutive model. I modified the xml originally posted in issue #1510 to do that. GEOSX fails with the following output:
```
Max threads: 72                                                                                                                                                                          
MKL max threads: 1                                                                                                                                                                       
Adding Solver of type CompositionalMultiphaseFVM, named compflow                                                                                                                         
Adding Mesh: InternalMesh, mesh                                                                                                                                                          
Adding Geometric Object: Box, source                                                                                                                                                     
Adding Geometric Object: Box, sink                                                                                                                                                       
Adding Event: PeriodicEvent, outputs                                                                                                                                                     
Adding Event: PeriodicEvent, solverApplications                                                                                                                                          
Adding Event: PeriodicEvent, timeHistoryOutput                                                                                                                                           
Adding Event: PeriodicEvent, timeHistoryCollection                                                                                                                                       
Adding Output: VTK, vtkOutput                                                                                                                                                            
Adding Output: TimeHistory, timeHistoryOutput                                                                                                                                            
Adding Object CellElementRegion named region from ObjectManager::Catalog.                                                                                                                
   TableFunction: PVDG_Bg                                                                                                                                                                
   TableFunction: PVDG_viscg                                                                                                                                                             
  region/cellBlock/fluid is allocated with 1 quadrature points.                                                                                                                          
  region/cellBlock/rock is allocated with 1 quadrature points.                                                                                                                           
  region/cellBlock/relperm is allocated with 1 quadrature points.                                                                                                                        
  region/cellBlock/rockPerm is allocated with 1 quadrature points.                                                                                                                       
  region/cellBlock/rockPorosity is allocated with 1 quadrature points.                                                                                                                   
  region/cellBlock/nullSolid is allocated with 1 quadrature points.                                                                                                                      
Time: 0s, dt:0.001s, Cycle: 0                                                                                                                                                            
*** Error in `/g/g90/hamon1/geosx/GEOSX/build-quartz-clang@10.0.0-release/bin/geosx': double free or corruption (out): 0x0000000004153030 ***                                            
slurmstepd: error: *** STEP 7326204.0 ON quartz1 CANCELLED AT 2021-08-30T09:45:07 ***                                                                                                    
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.                                                                                                                 
srun: error: quartz1: task 0: Segmentation fault (core dumped)                                                                                                                           
```

**To Reproduce**
1. Checkout `develop`
2. Run the xml attached: 
[issue.zip](https://github.com/GEOSX/GEOSX/files/7078300/issue.zip)

**Expected behavior**
I think it should work.

**Platform:**
- Platform: Quartz
- Compiler: clang 10.0.0 release
- GEOSX with the current `develop`

**Additional context**
It seems to me that that this problem is different from the original problem solved in #1510. I see that the function `resizeFields` of the constitutive model is called before the initialization of `PackCollection`, so the sizes should be known when `PackCollection` set up?  @wrtobin do you know what may be going on?

**Comment by wrtobin** [comment]
@francoishamon 

I'm attempting to recreate this on the current develop head (a8711cad3) on quartz, using the default debug config, and I'm getting an error from HYPRE, regardless of whether the run is serial or parallel (across x, the only dim with multiple elements), and with/without the time-history output commented out of the xml input.

Have you tried invoking the xml recently without the time history collection/output in the input deck?

```
Max threads: 18
MKL max threads: 18
GEOSX version 0.2.0 (bugfix/wrtobin/time-hist-issues, sha1: a8711cad3)
Adding Solver of type CompositionalMultiphaseFVM, named compflow
Adding Mesh: InternalMesh, mesh
Adding Geometric Object: Box, source
Adding Geometric Object: Box, sink
Adding Event: PeriodicEvent, outputs
Adding Event: PeriodicEvent, solverApplications
Adding Event: PeriodicEvent, timeHistoryOutput
Adding Event: PeriodicEvent, timeHistoryCollection
Adding Output: VTK, vtkOutput
Adding Output: TimeHistory, timeHistoryOutput
Adding Object CellElementRegion named region from ObjectManager::Catalog.
   TableFunction: fluid_PVDG_Bg
   TableFunction: fluid_PVDG_viscg
  region/cellBlock/fluid is allocated with 1 quadrature points.
  region/cellBlock/rock is allocated with 1 quadrature points.
  region/cellBlock/relperm is allocated with 1 quadrature points.
  region/cellBlock/rockPerm is allocated with 1 quadrature points.
  region/cellBlock/rockPorosity is allocated with 1 quadrature points.
  region/cellBlock/nullSolid is allocated with 1 quadrature points.
Time: 0s, dt:0.001s, Cycle: 0
Received signal 8: Floating point exception
Floating point exception:

** StackTrace of 24 frames **
Frame 0: LvArray::system::stackTraceHandler(int, bool) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/LvArray/src/system.cpp:478
Frame 1: /usr/WS1/tobin6/geosx/bugfix/input_decks/time-hist-issues/../../time-hist-issues/build-quartz-clang@10.0.0-debug/bin/geosx /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/common/initializeEnvironment.cpp:82
Frame 2: /usr/WS1/tobin6/geosx/bugfix/input_decks/time-hist-issues/../../time-hist-issues/build-quartz-clang@10.0.0-debug/bin/geosx /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/common/initializeEnvironment.cpp:82
Frame 3: /lib64/libc.so.6 
Frame 4: /usr/WS1/tobin6/geosx/bugfix/input_decks/time-hist-issues/../../time-hist-issues/build-quartz-clang@10.0.0-debug/bin/geosx csr_matvec.c:?
Frame 5: __kmp_invoke_microtask 
Frame 6: /usr/tce/packages/clang/clang-10.0.0/release/lib/libomp.so 
Frame 7: __kmp_fork_call 
Frame 8: __kmpc_fork_call 
Frame 9: hypre_CSRMatrixMatvecOutOfPlaceHost 
Frame 10: hypre_ParCSRMatrixMatvecOutOfPlace 
Frame 11: geosx::HypreMatrix::gemv(double, geosx::HypreVector const&, double, geosx::HypreVector&, bool) const /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/linearAlgebra/interfaces/hypre/HypreMatrix.cpp:768
Frame 12: geosx::MatrixBase<geosx::HypreMatrix, geosx::HypreVector>::residual(geosx::HypreVector const&, geosx::HypreVector const&, geosx::HypreVector&) const /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/linearAlgebra/interfaces/MatrixBase.hpp:526
Frame 13: geosx::SuiteSparse<geosx::HypreInterface>::solve(geosx::HypreVector const&, geosx::HypreVector&) const /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:262
Frame 14: geosx::SolverBase::solveSystem(geosx::DofManager const&, geosx::HypreMatrix&, geosx::HypreVector&, geosx::HypreVector&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/physicsSolvers/SolverBase.cpp:823
Frame 15: geosx::CompositionalMultiphaseBase::solveSystem(geosx::DofManager const&, geosx::HypreMatrix&, geosx::HypreVector&, geosx::HypreVector&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:1066
Frame 16: geosx::SolverBase::nonlinearImplicitStep(double const&, double const&, int, geosx::DomainPartition&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/physicsSolvers/SolverBase.cpp:588
Frame 17: geosx::CompositionalMultiphaseBase::solverStep(double const&, double const&, int, geosx::DomainPartition&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp:574
Frame 18: geosx::SolverBase::execute(double, double, int, int, double, geosx::DomainPartition&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/physicsSolvers/SolverBase.cpp:163
Frame 19: geosx::EventBase::execute(double, double, int, int, double, geosx::DomainPartition&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/events/EventBase.cpp:226
Frame 20: geosx::EventManager::run(geosx::DomainPartition&) /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/events/EventManager.cpp:173
Frame 21: geosx::ProblemManager::runSimulation() /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/mainInterface/ProblemManager.cpp:713
Frame 22: geosx::GeosxState::run() /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/coreComponents/mainInterface/GeosxState.cpp:203
Frame 23: main /usr/WS1/tobin6/geosx/bugfix/time-hist-issues/src/main/main.cpp:51
Frame 24: __libc_start_main 
=====
```

**Comment by francoishamon** [comment]
Hi @wrtobin 
Sorry to waste your time with this, I don't know what happened in this test case. While I am fixing it, I reproduced the error on an integrated test to make things easier (see attached).

```
Max threads: 72                                                                                                                                                                                                 
MKL max threads: 1                                                                                                                                                                                              
GEOSX version 0.2.0 (develop-cea27cd0a)                                                                                                                                                                         
Adding Solver of type CompositionalMultiphaseFVM, named compflow                                                                                                                                                
Adding Mesh: InternalMesh, mesh1                                                                                                                                                                                
Adding Geometric Object: Box, source                                                                                                                                                                            
Adding Geometric Object: Box, sink1                                                                                                                                                                             
Adding Geometric Object: Box, sink2                                                                                                                                                                             
Adding Geometric Object: Box, barrier                                                                                                                                                                           
Adding Event: PeriodicEvent, solverApplications                                                                                                                                                                 
Adding Event: PeriodicEvent, outputs                                                                                                                                                                            
Adding Event: PeriodicEvent, timeHistoryOutput                                                                                                                                                                  
Adding Event: PeriodicEvent, timeHistoryCollection                                                                                                                                                              
Adding Event: PeriodicEvent, restarts                                                                                                                                                                           
Adding Output: VTK, vtkOutput                                                                                                                                                                                   
Adding Output: TimeHistory, timeHistoryOutput                                                                                                                                                                   
Adding Output: Restart, sidreRestart                                                                                                                                                                            
Adding Object CellElementRegion named Region1 from ObjectManager::Catalog.                                                                                                                                      
   TableFunction: fluid1_SpanWagnerCO2Density_table                                                                                                                                                             
   TableFunction: fluid1_FenghourCO2Viscosity_table                                                                                                                                                             
   TableFunction: fluid1_BrineCO2Density_table                                                                                                                                                                  
   TableFunction: fluid1_CO2Solubility_table                                                                                                                                                                    
  Region1/cb1/fluid1 is allocated with 1 quadrature points.                                                                                                                                                     
  Region1/cb1/rock is allocated with 1 quadrature points.                                                                                                                                                       
  Region1/cb1/relperm is allocated with 1 quadrature points.                                                                                                                                                    
  Region1/cb1/rockPerm is allocated with 1 quadrature points.                                                                                                                                                   
  Region1/cb1/rockPorosity is allocated with 1 quadrature points.                                                                                                                                               
  Region1/cb1/nullSolid is allocated with 1 quadrature points.                                                                                                                                                  
Time: 0s, dt:50000s, Cycle: 0
    Attempt:  0, NewtonIter:  0                                                                                                                                                                                 
    ( Rfluid ) = (8.62e-01) ;     ( R ) = ( 8.62e-01 ) ;                                                                                                                                                        
    Attempt:  0, NewtonIter:  1                                                                                                                                                                                 
    ( Rfluid ) = (3.84e-01) ;     ( R ) = ( 3.84e-01 ) ;                                                                                                                                                        
    Last LinSolve(iter,res) = (   1, 3.96e-15 ) ;                                                                                                                                                               
    Attempt:  0, NewtonIter:  2                                                                                                                                                                                 
    ( Rfluid ) = (7.58e-02) ;     ( R ) = ( 7.58e-02 ) ;                                                                                                                                                        
    Last LinSolve(iter,res) = (   1, 7.60e-16 ) ;                                                                                                                                                               
    Attempt:  0, NewtonIter:  3                                                                                                                                                                                 
    ( Rfluid ) = (3.01e-02) ;     ( R ) = ( 3.01e-02 ) ;                                                                                                                                                        
    Last LinSolve(iter,res) = (   1, 3.96e-16 ) ;                                                                                                                                                               
    Attempt:  0, NewtonIter:  4                                                                                                                                                                                 
    ( Rfluid ) = (1.42e-03) ;     ( R ) = ( 1.42e-03 ) ;                                                                                                                                                        
    Last LinSolve(iter,res) = (   1, 2.80e-15 ) ;                                                                                                                                                               
    Attempt:  0, NewtonIter:  5                                                                                                                                                                                 
    ( Rfluid ) = (5.22e-06) ;     ( R ) = ( 5.22e-06 ) ;                                                                                                                                                        
    Last LinSolve(iter,res) = (   1, 1.14e-15 ) ;                                                                                                                                                               
compflow: Max phase CFL number: 2.56187                                                                                                                                                                         
compflow: Max component CFL number: 1.26349                                                                                                                                                                     
compflow: Newton solver converged in less than 40 iterations, time-step required will be doubled.                                                                                                               
*** Error in `./bin/geosx': munmap_chunk(): invalid pointer: 0x0000000003fb84e0 ***                                                                                                                             
slurmstepd: error: *** STEP 7411282.0 ON quartz16 CANCELLED AT 2021-09-22T16:48:50 ***                                                                                                                          
srun: Job step aborted: Waiting up to 62 seconds for job step to finish.                                                                                                                                        
srun: error: quartz16: task 0: Segmentation fault (core dumped)            
```
In the input file, if you switch from `fluid1_phaseMassDensity` to `phaseVolumeFraction` (not stored in a constitutive model), then everything works well. That is why I am thinking that the problem may come from the fact that the quantity lives in a constitutive model. Let me know if I can help, and thanks for looking into this.
[issue.zip](https://github.com/GEOSX/GEOSX/files/7214479/issue.zip)

**Comment by wrtobin** [comment]
@francoishamon In trying to recreate this on my branch which includes fixes related to 2 other issues, I am unable to reproduce the failure, and the output is generated as expected. This issue is also likely resolved by the fix developed for #1597 in particular.

**Comment by francoishamon** [comment]
@wrtobin That's great news, thanks!
